src/pnpq/units.py

```diff
@@ -2,5 +2,33 @@
 
 ureg = pint.UnitRegistry()
 
-# Custom unit definitions
-ureg.define("mpc320_step = (170 / 1370) degree")  # MPC320
+# Initialize the UnitRegistry
+ureg = pint.UnitRegistry()
+
+# Custom unit definitions for MPC320
+ureg.define("mpc320_step = (170 / 1370) degree")
+ureg.define("mpc320_max_velocity = 400 degree / second")
+
+# According to the protocol, velocity is expressed as a percentage of the maximum speed, ranging from 10% to 100%.
+# The maximum velocity is defined as 400 degrees per second, so we store velocity as a dimensionless proportion of this value.
+# Thus, the unit for mpc_velocity will be set as dimensionless.
+# A transformation function (defined below) will convert other units, like degrees per second, into this proportional form.
+ureg.define("mpc320_velocity = []")
+
+
+context = pint.Context("mpc320_proportional_velocity")
+
+context.add_transformation(
+    "degree / second",
+    "mpc320_velocity",
+    lambda ureg, value, **kwargs: value / ureg.mpc320_max_velocity * 100,
+)
```

Comment:
Hint: the syntax here looks a bit different from [the syntax in the manual](https://pint.readthedocs.io/en/stable/user/contexts.html#defining-contexts-programmatically).

Reply:
Also, let's do the rounding here. We discussed on Friday how that might lead to a loss of accuracy, but that's not really a concern here: any time we are using the mpc320_velocity unit, we want to know what the velocity of the MPC320 is. We don't need to know what the person who sent the command _wanted_ the velocity to be, if they sent a velocity that doesn't translate to a value we can send to the MPC320.

We can also throw an exception if the velocity is out of bounds!

This might sound like a lot to do in a lambda, but `lambda` is just a shorthand for defining a function. You can define a normal, multi-line function instead, and then just pass the name of the function here.

Reply:
> Can you write additional unit tests to check if this also works with radians? If not, is there a way we can generalize this to work with all types of angle measurement... even mpc320 steps?

It should automatically work with other types of `[ angle ]`, because pint automatically will convert it to degrees before doing the calculation. So, it should also work with `mpc320_step` as it's a variant of angle. I'll add some unit tests for radians and `mpc320_step` as well.

Reply:
> Also, let's do the rounding here. We discussed on Friday how that might lead to a loss of accuracy, but that's not really a concern here: any time we are using the mpc320_velocity unit, we want to know what the velocity of the MPC320 is. We don't need to know what the person who sent the command _wanted_ the velocity to be, if they sent a velocity that doesn't translate to a value we can send to the MPC320.
> 
> We can also throw an exception if the velocity is out of bounds!
> 
> This might sound like a lot to do in a lambda, but `lambda` is just a shorthand for defining a function. You can define a normal, multi-line function instead, and then just pass the name of the function here.

Okay, I've been trying to make rounding work for a while now, but it doesn't seem possible with this `Context` method. Currently, the transformation doesn't round, but when I try to add it like so,
```py
context.add_transformation(
    "degree / second",
    "mpc320_velocity",
    lambda ureg, value, **kwargs: round((value / mpc320_max_velocity) * 100),
)
```
All forms of angle conversion simply return zero. e.g:
```py
print((100 * ureg("degree / sec").to("mpc320_velocity"))) # prints 0.0 mpc320_velocity
```
Trying to make a custom function like you suggested:
```py
def to_mpc320_velocity(ureg, value, **kwargs) -> Quantity:
    converted_velocity = (value / mpc320_max_velocity) * 100 
    print(converted_velocity)
    return round(converted_velocity)
``` 
also has the same issue. But, the print value in the function always seems to print `0.25 mpc320_velocity` no matter what number I put for `value`. Since `0.25 mpc320_velocity == 1 degree / second`, my belief is that pint internally calculates the ratio for when the input is set to `1 degree / second` and then scales that accordingly.

So I'm not sure how to do rounding for now...

Reply:
It works for me. The only issue I can see is that the magnitude is still represented as a float rather than an int, which will potentially cause problems. I copied the module into another file to do a little testing:

```python
#!/usr/bin/env python

import pint

ureg = pint.UnitRegistry()

# Custom unit definitions
ureg.define("mpc320_step = (170 / 1370) degree")  # MPC320

ureg.define("mpc320_velocity = []")

context = pint.Context("mpc320_proportional_velocity")

mpc320_max_velocity = 400 * (ureg.degree / ureg.second)


context.add_transformation(
    "degree / second",
    "mpc320_velocity",
    lambda ureg, value: round((value / mpc320_max_velocity) * 100),  # Convert value to percent
)

context.add_transformation(
    "mpc320_velocity",
    "degree / second",
    lambda ureg, value: (value * mpc320_max_velocity) / 100,  # Convert value from percent
)

ureg.add_context(context)
ureg.enable_contexts("mpc320_proportional_velocity")

velocity = 50 * ureg.mpc320_velocity
print(velocity)

velocity = velocity.to("degree / second")
print(velocity)

velocity = velocity.to("mpc320_velocity")
print(velocity)

velocity = 201 * (ureg.degree / ureg.second)
print(velocity)

velocity = velocity.to("mpc320_velocity")
print(velocity)

velocity = velocity.to("degree / second")
print(velocity)
```

Reply:
Yes, that code works perfectly fine for me now. I am not sure what was up before. I was likely on a drug or perhaps two drugs.